File: autocontext/scripts/drive_parity_fixture.py
#!/usr/bin/env python3
"""
Cross-runtime parity fixture driver — Python runtime.

Usage: uv run python scripts/drive_parity_fixture.py <fixture-name>

Reads fixture inputs, runs instrument_client with a mock httpx transport,
captures the emitted trace, normalizes non-deterministic fields, and prints
canonical JSON to stdout.

Exit 0 on success, 1 on error.
"""
from __future__ import annotations

import gc
import json
import os
import sys
import tempfile
from pathlib import Path
from typing import Any

# Add the autocontext src to path
ROOT = Path(__file__).parent.parent
FIXTURES_DIR = (
    ROOT.parent
    / "ts"
    / "tests"
    / "integrations"
    / "openai"
    / "parity"
    / "fixtures"
)


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python drive_parity_fixture.py <fixture-name>", file=sys.stderr)
        sys.exit(1)

    fixture_name = sys.argv[1]
    fixture_dir = FIXTURES_DIR / fixture_name
    if not fixture_dir.exists():
        print(f"Fixture not found: {fixture_dir}", file=sys.stderr)
        sys.exit(1)

    request_json = json.loads((fixture_dir / "request.json").read_text())
    identity_json = json.loads((fixture_dir / "identity.json").read_text())
    is_error = (fixture_dir / "error.json").exists()
    is_streaming = request_json.get("stream", False)
    is_responses_api = "input" in request_json or request_json.get("endpoint") == "responses"

    import httpx

    # Build mock transport
    if is_error:
        error_json = json.loads((fixture_dir / "error.json").read_text())

        _err_data = error_json
        def _error_handler(request: httpx.Request) -> httpx.Response:
            return httpx.Response(
                status_code=_err_data["status"],
                json={"error": {"message": _err_data["message"], "type": "api_error", "code": None}},
            )
        transport = httpx.MockTransport(_error_handler)
    elif is_streaming:
        chunks = json.loads((fixture_dir / "response.json").read_text())

        def _stream_handler(request: httpx.Request) -> httpx.Response:
            lines = ""
            for chunk in chunks:
                lines += f"data: {json.dumps(chunk)}\n\n"
            lines += "data: [DONE]\n\n"
            return httpx.Response(
                status_code=200,
                content=lines.encode("utf-8"),
                headers={"content-type": "text/event-stream"},
            )
        transport = httpx.MockTransport(_stream_handler)
    else:
        response_json = json.loads((fixture_dir / "response.json").read_text())

        def _json_handler(request: httpx.Request) -> httpx.Response:
            return httpx.Response(status_code=200, json=response_json)
        transport = httpx.MockTransport(_json_handler)

    from openai import OpenAI
    from autocontext.integrations.openai import FileSink, instrument_client, autocontext_session

    with tempfile.TemporaryDirectory() as tmp_dir:
        trace_path = Path(tmp_dir) / "traces.jsonl"
        sink = FileSink(trace_path, batch_size=1, flush_interval_seconds=0)

        # Handle salt for session fixtures
        salt_file = fixture_dir / "install-salt.txt"
        original_dir = os.getcwd()
        if salt_file.exists():
            # Write salt to a .autocontext/install-salt path relative to fixture_dir
            salt_dir = Path(tmp_dir)
            (salt_dir / ".autocontext").mkdir(exist_ok=True)
            (salt_dir / ".autocontext" / "install-salt").write_text(salt_file.read_text().strip())
            os.chdir(salt_dir)

        try:
            inner = OpenAI(api_key="test-key", http_client=httpx.Client(transport=transport), max_retries=0)
            client = instrument_client(inner, sink=sink, app_id="parity-test-app", environment_tag="test")

            def run_request() -> None:
                if is_responses_api:
                    try:
                        client.responses.create(**request_json)
                    except Exception:
                        pass
                elif is_streaming:
                    try:
                        stream = client.chat.completions.create(**request_json)
                        if fixture_name == "chat-streaming-abandoned":
                            # Read first chunk then abandon
                            it = iter(stream)
                            next(it)
                            del stream
                            del it
                            gc.collect()
                        else:
                            for _chunk in stream:
                                pass
                    except Exception:
                        pass
                else:
                    try:
                        client.chat.completions.create(**request_json)
                    except Exception:
                        pass
                sink.flush()
                sink.close()

            if identity_json.get("userId") or identity_json.get("sessionId"):
                with autocontext_session(
                    user_id=identity_json.get("userId"),
                    session_id=identity_json.get("sessionId"),
                ):
                    run_request()
            else:
                run_request()

            # Read the emitted trace
            content = trace_path.read_text().strip()
            if not content:
                print("No trace emitted", file=sys.stderr)
                sys.exit(1)

            raw_trace = json.loads(content.split("\n")[0])
        finally:
            os.chdir(original_dir)

    # Normalize non-deterministic fields
    normalized = normalize_trace(raw_trace, fixture_name)

    # Print canonical JSON
    print(canonical_json(normalized))


def normalize_trace(trace: dict[str, Any], fixture_name: str) -> dict[str, Any]:
    t = dict(trace)
    # Normalize traceId
    t["traceId"] = "PARITY_TRACE_ID_NORMALIZED"
    # Normalize timing
    t["timing"] = {
        "startedAt": "2024-01-01T00:00:00Z",
        "endedAt": "2024-01-01T00:00:01Z",
        "latencyMs": 1000,
    }
    # Normalize SDK name + version (different runtimes have different names)
    if isinstance(t.get("source"), dict) and isinstance(t["source"].get("sdk"), dict):
        t["source"] = dict(t["source"])
        t["source"]["sdk"] = {"name": "autocontext-sdk", "version": "0.0.0"}
    # Normalize message timestamps
    if isinstance(t.get("messages"), list):
        t["messages"] = [
            {**m, "timestamp": "2024-01-01T00:00:00Z"}
            for m in t["messages"]
        ]
    # Normalize error fields (message format, stack, and error-type vary between SDK versions/runtimes)
    if isinstance(t.get("outcome"), dict) and isinstance(t["outcome"].get("error"), dict):
        t["outcome"] = dict(t["outcome"])
        err = dict(t["outcome"]["error"])
        if "stack" in err:
            err["stack"] = "NORMALIZED"
        if "message" in err:
            err["message"] = "NORMALIZED"
        if "type" in err:
            err["type"] = "NORMALIZED"
        t["outcome"]["error"] = err
    return t


def canonical_json(obj: Any) -> str:
    if isinstance(obj, list):
        return "[" + ",".join(canonical_json(v) for v in obj) + "]"
    if isinstance(obj, dict):
        keys = sorted(obj.keys())
        return "{" + ",".join(
            json.dumps(k) + ":" + canonical_json(obj[k]) for k in keys
        ) + "}"
    if obj is None:
        return "null"
    return json.dumps(obj)


if __name__ == "__main__":
    main()

File: ts/scripts/drive-parity-fixture.mjs
#!/usr/bin/env node
// Run with: node --import tsx/esm scripts/drive-parity-fixture.mjs
/**
 * Cross-runtime parity fixture driver — TypeScript runtime.
 *
 * Usage: node scripts/drive-parity-fixture.mjs <fixture-name>
 *
 * Reads fixture inputs, runs instrumentClient with a mock OpenAI client,
 * captures the emitted trace, normalizes non-deterministic fields (traceId,
 * timestamps, latencyMs, SDK version), and prints canonical JSON to stdout.
 *
 * Exit 0 on success, 1 on error.
 */

import { readFileSync, existsSync, mkdtempSync, rmSync } from "node:fs";
import { join, dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { tmpdir } from "node:os";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(__dirname, "..");
const FIXTURES_DIR = join(ROOT, "tests", "integrations", "openai", "parity", "fixtures");

const fixtureName = process.argv[2];
if (!fixtureName) {
  process.stderr.write("Usage: node drive-parity-fixture.mjs <fixture-name>\n");
  process.exit(1);
}

const fixtureDir = join(FIXTURES_DIR, fixtureName);
if (!existsSync(fixtureDir)) {
  process.stderr.write(`Fixture not found: ${fixtureDir}\n`);
  process.exit(1);
}

// Load fixture files
const requestJson = JSON.parse(readFileSync(join(fixtureDir, "request.json"), "utf-8"));
const identityJson = JSON.parse(readFileSync(join(fixtureDir, "identity.json"), "utf-8"));
const isError = existsSync(join(fixtureDir, "error.json"));
const isStreaming = requestJson.stream === true;
const isResponsesApi = ("input" in requestJson || requestJson.endpoint === "responses");

// Set up a temp sink
const tmpDir = mkdtempSync(join(tmpdir(), "parity-driver-"));
const tracePath = join(tmpDir, "traces.jsonl");

try {
  // Dynamic import to get ESM modules from src
  const { FileSink } = await import("../src/integrations/openai/sink.js");
  const { instrumentClient } = await import("../src/integrations/openai/wrap.js");
  const { autocontextSession } = await import("../src/integrations/openai/session.js");

  // Build mock fetch
  let mockFetch;
  if (isError) {
    const errorJson = JSON.parse(readFileSync(join(fixtureDir, "error.json"), "utf-8"));
    mockFetch = (_url, _init) => Promise.resolve(
      new Response(
        JSON.stringify({ error: { message: errorJson.message, type: "api_error", code: null } }),
        { status: errorJson.status, headers: { "content-type": "application/json" } },
      ),
    );
  } else if (isStreaming) {
    const chunks = JSON.parse(readFileSync(join(fixtureDir, "response.json"), "utf-8"));
    mockFetch = (_url, _init) => {
      const lines = chunks.map(c => `data: ${JSON.stringify(c)}\n\n`).join("") + "data: [DONE]\n\n";
      return Promise.resolve(new Response(lines, {
        status: 200,
        headers: { "content-type": "text/event-stream" },
      }));
    };
  } else {
    const responseJson = JSON.parse(readFileSync(join(fixtureDir, "response.json"), "utf-8"));
    mockFetch = (_url, _init) => Promise.resolve(
      new Response(JSON.stringify(responseJson), {
        status: 200,
        headers: { "content-type": "application/json" },
      }),
    );
  }

  // Set up install salt for session fixtures
  const originalDir = process.cwd();
  const saltFile = join(fixtureDir, "install-salt.txt");
  let changedDir = false;
  if (existsSync(saltFile)) {
    // Create a temp dir with .autocontext/install-salt matching the fixture salt
    const { mkdirSync, writeFileSync, mkdtempSync: mkdtempSyncFs } = await import("node:fs");
    const saltTmpDir = mkdtempSyncFs(join(tmpdir(), "parity-salt-"));
    mkdirSync(join(saltTmpDir, ".autocontext"), { recursive: true });
    const saltContent = readFileSync(saltFile, "utf-8").trim();
    writeFileSync(join(saltTmpDir, ".autocontext", "install-salt"), saltContent);
    process.chdir(saltTmpDir);
    changedDir = true;
  }

  // Create OpenAI-like client mock
  const { default: OpenAI } = await import("openai");
  const inner = new OpenAI({ apiKey: "test-key", fetch: mockFetch, maxRetries: 0 });

  const sink = new FileSink(tracePath, { batchSize: 1, flushIntervalSeconds: 0 });
  const client = instrumentClient(inner, {
    sink,
    appId: "parity-test-app",
    environmentTag: "test",
  });

  // Handle session identity
  const runRequest = async () => {
    // Run the request
    if (isResponsesApi) {
      try {
        await client.responses.create(requestJson);
      } catch {
        // expected for error fixtures
      }
    } else if (isStreaming) {
      try {
        if (fixtureName === "chat-streaming-abandoned") {
          // Wrap in a sub-function so stream + iter go out of scope when it returns,
          // making them eligible for GC before we call gc().
          await (async () => {
            const stream = await client.chat.completions.create(requestJson);
            const iter = stream[Symbol.asyncIterator]();
            await iter.next(); // read one chunk, then let everything go out of scope
            // Do NOT reference stream or iter after this point
          })();
          // Now stream and iter are out of scope; force GC if available
          if (typeof gc !== "undefined") {
            gc();
            gc(); // second pass to collect cycles
          }
          // Wait a bit for FinalizationRegistry callbacks to fire
          await new Promise(r => setTimeout(r, 200));
        } else {
          const stream = await client.chat.completions.create(requestJson);
          for await (const _chunk of stream) { /* consume */ }
        }
      } catch {
        // expected for error fixtures
      }
    } else {
      try {
        await client.chat.completions.create(requestJson);
      } catch {
        // expected for error fixtures
      }
    }
    sink.flush();
    sink.close();
  };

  if (identityJson.userId || identityJson.sessionId) {
    await autocontextSession(
      { userId: identityJson.userId, sessionId: identityJson.sessionId },
      runRequest,
    );
  } else {
    await runRequest();
  }

  // Restore original directory if we changed it
  if (changedDir) {
    process.chdir(originalDir);
  }

  // Read the emitted trace
  let rawTrace;
  try {
    const content = readFileSync(tracePath, "utf-8").trim();
    if (!content) {
      process.stderr.write("No trace emitted\n");
      process.exit(1);
    }
    rawTrace = JSON.parse(content.split("\n")[0]);
  } catch (e) {
    process.stderr.write(`Failed to read trace: ${e}\n`);
    process.exit(1);
  }

  // Normalize non-deterministic fields for cross-runtime parity
  const normalized = normalizeTrace(rawTrace, fixtureName);

  // Print canonical JSON (sorted keys, no spaces)
  process.stdout.write(canonicalJson(normalized) + "\n");
  process.exit(0);
} finally {
  rmSync(tmpDir, { recursive: true, force: true });
}

function normalizeTrace(trace, fixtureName) {
  const t = { ...trace };
  // Normalize traceId → deterministic constant
  t.traceId = "PARITY_TRACE_ID_NORMALIZED";
  // Normalize timing
  t.timing = {
    startedAt: "2024-01-01T00:00:00Z",
    endedAt: "2024-01-01T00:00:01Z",
    latencyMs: 1000,
  };
  // Normalize SDK name + version in source (different runtimes have different names)
  if (t.source?.sdk) {
    t.source = { ...t.source, sdk: { name: "autocontext-sdk", version: "0.0.0" } };
  }
  // Normalize message timestamps
  if (Array.isArray(t.messages)) {
    t.messages = t.messages.map(m => ({ ...m, timestamp: "2024-01-01T00:00:00Z" }));
  }
  // Normalize error fields (message format, stack, and error-type vary between SDK versions/runtimes)
  if (t.outcome?.error) {
    const err = { ...t.outcome.error };
    if (err.stack) err.stack = "NORMALIZED";
    if (err.message) err.message = "NORMALIZED";
    if (err.type) err.type = "NORMALIZED";
    t.outcome = { ...t.outcome, error: err };
  }
  return t;
}

function canonicalJson(obj) {
  if (Array.isArray(obj)) return "[" + obj.map(canonicalJson).join(",") + "]";
  if (obj === null) return "null";
  if (typeof obj !== "object") return JSON.stringify(obj);
  const keys = Object.keys(obj).sort();
  return "{" + keys.map(k => JSON.stringify(k) + ":" + canonicalJson(obj[k])).join(",") + "}";
}
